README.md

```diff
@@ -18,4 +20,167 @@ Execute the following steps to practice various topics covered in the intro to g
+
+```bash
+git log
+```
+
+However, the output of the above command may be too verbose. You can use the following command to see a more concise output.
+
+```bash
+git log --all --decorate --oneline --graph
+```
+
+In the current state, you should see a commit history that looks like this:
+
+![Commit Log](/content/images/log_mid.png)
+
+Of course, the commit hashes and messages might be different in your case.
+
+Feel free to use the above command to visualize the commits that you are making.
+
+## Rebase
+
+`git rebase` is used to reapply commits on top of another base tip. This is useful to maintain a linear project history.
+
+Do the following steps for a rebase:
+
+1) Create a new branch called `mySecondBranch` (try to recall how to create a new branch). And don't forget to switch to the new branch.
+2) Add a new text file called `new2.txt` and add some text to it. Add and commit these changes.
+3) Similarly, add a new text file called `new3.txt` and add some text to it. Add and commit these changes.
+4) Now, switch back to the `main` branch and add another text file called `new4.txt` and add some text to it. Add and commit these changes.
+
+   If you have followed the steps correctly, you should have a commit history that looks like this:
+
+    ![Commit Log 1](/content/images/log_rebase_1.png)
+
+    You can verify this by using the `git log` command.
+5) Now, rebase `mySecondBranch` onto `main` using the following commands:
+
+    ```bash
+    git checkout mySecondBranch
+    git rebase main
+    ```
+
+6) Verify that your commit history now looks like this:
+
+    ![Commit Log 2](/content/images/log_rebase_2.png)
+
+    You can use the `git log` command to verify this.
+
+    Notice that the commit history is now linear, and the commits from `mySecondBranch` have been reapplied on top of the `main` branch. However, the commit hashes will be different before and after the rebase.
+
+### Rebase vs Merge
+
+In cases where we want to combine the changes from one branch to another, we can use either `rebase` or `merge`.
+
+| Merge                                                            | Rebase                                                          |
+| ---------------------------------------------------------------- | -------------------------------------------------------------- |
+| Creates a new commit that combines the changes from two branches | Reapplies the commits from one branch on top of another branch |
+| Creates a non-linear commit history                              | Creates a linear commit history                                |
+| Preserves the commit history of both branches                    | Rewrites the commit history of the branch being rebased |
+
+Now that you have rebased the `mySecondBranch` branch on top of the `main` branch, use `git log` to compare the commit history with the one where you merged `myFirstBranch` with `main`.
+
```

Comment:
Better to have picture here as well. Do you use VSCode? there is a remarkable feature that shows you the commit graph. Something like this:
![image](https://github.com/Programming-Club-IITM/GitPractice/assets/73378019/4c5afa84-90b0-4c4b-829c-edf603d833be)
Including that here would help. Or including a general picture from the internet would also be fine, one that shows the difference between rebasing and merging clearly

Reply:
yes i do have vscode, but do not know about such a feature

I came accross `gitk` as well, but I am not sure whether all environments have it. Nevertheless, it isn't very pretty.

In this case however, the merge of `myFirstBranch` would be a fast-forward merge. So I couldn't get anything to visualize it like a three-way-merge.

Reply:
I think it comes with gitlens/gitkraken idk which one exactly. Try to get it, super helpful. Put whatever pic seems fine. But include one. I would suggest getting this and then trying it out

Reply:
wow, its been 4 minutes since i installed gitlens and it is already quite impressive...
and yeah, i will add the necessary image